api/client/charms/client_test.go

```diff
@@ -424,3 +425,86 @@ func (s *charmsMockSuite) TestZipHasNoHooksNorDispath(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(hasHooks, jc.IsFalse)
 }
+
+// TestZipHasSingleHook tests that an archive containing only a single hook
+// file (and no zip entry for the hooks directory) is still validated as a
+// charm with hooks.
+func (s *charmsMockSuite) TestZipHasSingleHook(c *gc.C) {
+	tempFile, err := os.CreateTemp(c.MkDir(), "charm")
+	c.Assert(err, jc.ErrorIsNil)
+	defer tempFile.Close()
+
+	zipWriter := zip.NewWriter(tempFile)
+	// add a single install hook
+	_, err = zipWriter.Create("hooks/install")
+	c.Assert(err, jc.ErrorIsNil)
+	err = zipWriter.Close()
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Verify created zip is as expected
+	zipReader, err := zip.OpenReader(tempFile.Name())
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(len(zipReader.File), gc.Equals, 1)
+	c.Assert(zipReader.File[0].Name, gc.Equals, "hooks/install")
+	c.Assert(zipReader.File[0].Mode().IsRegular(), jc.IsTrue)
+
+	// Verify this is validated as having a hook
+	hasHooks, err := (*charms.HasHooksOrDispatch)(tempFile.Name())
+	c.Check(err, jc.ErrorIsNil)
+	c.Check(hasHooks, jc.IsTrue)
+}
+
+// TestZipEmptyHookDir tests that an archive containing only an empty hooks
+// directory is not validated as a charm with hooks.
+func (s *charmsMockSuite) TestZipEmptyHookDir(c *gc.C) {
+	tempFile, err := os.CreateTemp(c.MkDir(), "charm")
+	c.Assert(err, jc.ErrorIsNil)
+	defer tempFile.Close()
+
+	zipWriter := zip.NewWriter(tempFile)
+	// add an empty hooks directory
+	_, err = zipWriter.Create("hooks/")
+	c.Assert(err, jc.ErrorIsNil)
+	err = zipWriter.Close()
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Verify created zip is as expected
+	zipReader, err := zip.OpenReader(tempFile.Name())
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(len(zipReader.File), gc.Equals, 1)
+	c.Assert(zipReader.File[0].Name, gc.Equals, "hooks/")
+	c.Assert(zipReader.File[0].Mode().IsDir(), jc.IsTrue)
+
+	// Verify this is validated as having no hooks
+	hasHooks, err := (*charms.HasHooksOrDispatch)(tempFile.Name())
+	c.Check(err, jc.ErrorIsNil)
+	c.Check(hasHooks, jc.IsFalse)
+}
+
+// TestZipSubfileHook tests that an archive containing nested subfiles inside
+// the hooks directory (i.e. not in the top level) is not validated as a charm
+// with hooks.
+func (s *charmsMockSuite) TestZipSubfileHook(c *gc.C) {
+	tempFile, err := os.CreateTemp(c.MkDir(), "charm")
+	c.Assert(err, jc.ErrorIsNil)
+	defer tempFile.Close()
+
+	zipWriter := zip.NewWriter(tempFile)
+	// add a file inside a subdir of hooks
+	_, err = zipWriter.Create("hooks/foo/bar.sh")
```

Comment:
the only one that I'm not sure about is if you have:
"hooks/hooks/install"
or
"foo/hooks/install"

It might be worth adding those to this test.

Ideally all this code would be refactored so we don't have to create a zip file on disk to do our tests of shapes of directory structures, but I don't think that refactoring pays its weight in this case.

Reply:
Added those to the test, they were fine.